Add date range search for pages closes #81

SYU15 opened the pull request:
The code for this is rough af and I there are still a few bugs/edge cases I need to iron out but wanted your eyes on this early to take a look at the pages I am getting back as a response.

They don't seem to match the date ranges I am sending back (they seem to be pretty off, not just kinda off) so either my query is malformed or maybe there is an issue with the backend?

Also I had to add a separate `webpack` pipeline for `react-dates` (not sure if I should be excluding all `node_modules` in the other pipelines, might make it more explicit) because it does not support CSS modules, but it looks pretty 😭

Oh yeah, and will write tests once I have a clearer idea what the issues are with the code.

Mr0grog commented:
> They don't seem to match the date ranges I am sending back (they seem to be pretty off, not just kinda off) so either my query is malformed or maybe there is an issue with the backend?

Hmmm, I feel like I’m getting results that make sense, but I’m guessing this might be a misunderstanding of what this API call does.

`/pages?capture_time={timestamp1}..{timestamp2}` gets you all the pages that had *a version* captured during that timeframe, not that were *most recently captured* during that timeframe.

That’s probably exacerbated by the fact that the column labeled “capture date” is a bit of a misnomer — it’s the date of the latest capture: https://github.com/edgi-govdata-archiving/web-monitoring-ui/blob/d9138d9e5c7ae1a764228c01051ea123b67932ea/src/components/page-list.jsx#L78-L79

…which is not the thing the API call in question is looking at. (You are making the right API call; the way things are labeled here is what’s confusing.)

Mr0grog commented on src/components/web-monitoring-ui.jsx:
```diff
@@ -44,7 +46,8 @@ export default class WebMonitoringUi extends React.Component {
     this.loadPages = this.loadPages.bind(this);
-    this.search = this.search.bind(this);
+    this.urlSearch = this.urlSearch.bind(this);
+    this.dateSearch = this.dateSearch.bind(this);
```
This works, but the reason I had named the previous handler `search` and had it take a query *object* instead of just a URL was that I was thinking things would get way out of hand if we did something like this (after all, there are a lot of other fields we could search besides URL and date). `WebMonitoringUi.search()` was meant to take a query object with all the search fields set on it.

Mr0grog commented on src/components/web-monitoring-ui.jsx:
```diff
@@ -65,14 +68,29 @@ export default class WebMonitoringUi extends React.Component {
+            endDateQuery: this.state.endDateQuery
+          }
+        };
+      }
```
This formatting should really happen either before we set the state or inside `loadPages()` when we are formatting the arguments for `api.getPages()`. (Depends on whether you want to view the state as “current query args” or as “the data we have set in various fields.”)

Mr0grog commented on src/components/web-monitoring-ui.jsx:
```diff
@@ -65,14 +68,29 @@ export default class WebMonitoringUi extends React.Component {
+          }
+        };
+      }
+      this.loadPages(captureTime);
```
I don’t think `captureTime` should be special here. `loadPages()` needs to pull it out of the state like it does everything else, and not receive it as a special argument.

Mr0grog commented on src/services/web-monitoring-db.js:
```diff
@@ -287,6 +287,10 @@ export default class WebMonitoringDb {
         }
+        else if (key === CAPTURE_TIME) {
+          const formattedCaptureQuery = formatCaptureQuery(value);
+          queryList.push(formattedCaptureQuery);
```
I think it might be a good idea to separate out the formatting of the value from the formatting of the key here, like we do above for arrays.

Mr0grog commented on src/services/web-monitoring-db.js:
```diff
@@ -472,3 +478,21 @@ function throwIfError (summary) {
   };
 }
+
+function formatCaptureQuery (captureTime) {
```
A few thoughts here:

- You’re not handling the previously-used case where `captureTime` is already a formatted string, which is causing the page details view not to load. (Probably easiest to just treat this function as a pass-through for strings.)

- Looks like the date picker returns Moment.js instances, which is the only kind of input you’re handling here. We either:
    - Make sure to also handle `Date` instances or
    - Do an audit to make sure we only ever pass Moment instances in (so we never pass in, say, `someVersion.capture_time`, which is a `Date`).

    It might be easiest to split this up:

    ```js
    function formatDate (date) {
      if (!date) return '';
      else if (moment.isMoment(date)) return date.utc().toISOString();
      else if (date instanceof Date) return date.toISOString();
      else throw new TypeError(`formatDate only takes Moment or Date instances, not ${date.constructor.name}s`);
    }
    function formatDateRange(timeRange) {
      if (typeof timeRange === 'string') return timeRange;

      return `${formatDate(timeRange.startDate)}..${formatDate(timeRange.endDate)}`;
    }
    ```

SYU15 commented:
> …which is not the thing the API call in question is looking at. (You are making the right API call; the way things are labeled here is what’s confusing.)

Should we maybe add a label or description for the datepicker or something that makes it more descriptive? I can figure something out if so. I feel like if I was confused by the behavior, others would be too.

Mr0grog commented:
Sure. *Is* that the issue, that you misunderstood either what that column was showing or what the API was doing?

If we rename the column, I’m thinking something like “most recent change” or “latest change.”

SYU15 commented:
> If we rename the column, I’m thinking something like “most recent change” or “latest change.”

Yeah I think renaming the column is probably the most straightforward/understandable change tbh. Will work on that.

SYU15 commented on src/components/web-monitoring-ui.jsx:
```diff
@@ -65,11 +65,22 @@ export default class WebMonitoringUi extends React.Component {
     api.logOut();
     this.setState({user: api.userData});
-    this.loadPages('pages');
+    this.loadPages();
```
I want to refactor this to call `loadPages` on `componentDidUpdate` but it after looking at the code, it is not as simple as just putting that function inside the lifecycle hook. Will probably take this on in a later PR unless you have concerns

SYU15 commented:
Code getting better, tests still TBD (and need to be fixed, obvi). 

The main reason I wanted to push this up again is because I am experiencing (what I think) is unexpected behavior when I try to combine both `url` and `capture_time` in my search. I may be misunderstanding what the queries together mean. 

For example, when I search along a date range and I see one EPA page, my expectation is that if I add the EPA regex into the query, it will drill down into that one page, but instead, I am not getting results. Oddly enough, it seems if I search for the URL first and then select a date range, I get what I expected.

This URL does not work (but just querying for capture_time returns an EPA page, also tried with NASA with similar results):
https://api-staging.monitoring.envirodatagov.org/api/v0/pages?include_latest=true&url=*%2F%2Fepa*&capture_time=2019-09-01T12:00:00.000Z..2019-09-25T12:00:00.000Z

This URL does work (searched for URL first and then capture_time):
https://api-staging.monitoring.envirodatagov.org/api/v0/pages?include_latest=true&url=*%2F%2Fepa*&capture_time=2019-01-01T20:00:00.000Z

Am I going crazy here?

Mr0grog commented:
> if I add the EPA regex

I don’t think you’re actually making this mistake, but just to be clear, the search expression here is not a regular expression. It’s just `*` as a wildcard.

> This URL does not work (but just querying for capture_time returns an EPA page, also tried with NASA with similar results):
> https://api-staging.monitoring.envirodatagov.org/api/v0/pages?include_latest=true&url=*%2F%2Fepa*&capture_time=2019-09-01T12:00:00.000Z..2019-09-25T12:00:00.000Z

I’m 95% sure the problem here is that we don’t have many pages with URLs matching `*//epa*`, and none that match that in the given timeframe. We *do* have lots of pages matching `*//www.epa*`, though: https://api-staging.monitoring.envirodatagov.org/api/v0/pages?include_latest=true&url=*%2F%2Fwww.epa*&capture_time=2019-09-01T12:00:00.000Z..2019-09-25T12:00:00.000Z

FWIW, I usually use `*epa.gov/*` for this type of query.

> This URL does work (searched for URL first and then capture_time):
api-staging.monitoring.envirodatagov.org/api/v0/pages?include_latest=true&url=*%2F%2Fepa*&capture_time=2019-01-01T20:00:00.000Z

I don’t get any results for this URL. Any chance you made a typo or copy/paste error here? (I would guess the problem is the exact value instead of a range for capture time…?)

---

Probably irrelevant side note: for domain queries, you could also use the `domain:<xyz>` and `2l-domain:<xyz>` tags, e.g:

https://api-staging.monitoring.envirodatagov.org/api/v0/pages?include_latest=true&capture_time=2019-09-01T12:00:00.000Z..2019-09-25T12:00:00.000Z&tags[]=2l-domain:epa.gov

SYU15 commented:
Ok I think I knew all of that stuff (including the wildcard) at one point and then forgot all of it, my bad. I guess my question is: why not refactor the wildcard thingy to be flexible enough so that a search for `epa` would return results for both `//www.epa` and `//epa`?

It it because there are too many entries returned? I feel like that will be somewhat mitigated by adding date range + pagination. Or is there another constraint/nuance of the analyst workflow that I'm not aware of?

Also no idea why that second url was truncated, since I copied it directly from chrome devtools but it's not that important since I understand the root cause now.

SYU15 commented:
It would require just a tiny change to this line, no?
https://github.com/edgi-govdata-archiving/web-monitoring-ui/blob/master/src/components/search-bar/search-bar.jsx#L32

Mr0grog commented:
> I guess my question is: why not refactor the wildcard thingy to be flexible enough so that a search for `epa` would return results for both `//www.epa` and `//epa`?

I think this one is probably a much bigger discussion… URL queries currently boil down to a `LIKE` query in SQL because that’s safe and efficient: https://github.com/edgi-govdata-archiving/web-monitoring-db/blob/459c23211455966826b1470562e1ebe81294d431/app/controllers/api/v0/pages_controller.rb#L161-L166

Full-text search didn’t seem like a good fit for URLs, and it’s relatively easy to craft a malicious regex, so allowing regexes and keeping it safe is super hard. I figured that, until people had major issues with search, it was best to keep the API extra simple and do all the fancy stuff in the client, hence code like: https://github.com/edgi-govdata-archiving/web-monitoring-ui/blob/f0d70a1559b3df48e2b056040495b7ff027ba63d/src/components/search-bar/search-bar.jsx#L28-L38

The database also has a SURT-formatted `url_key` field that might be helpful for stuff like this, but you can’t currently query it: https://github.com/edgi-govdata-archiving/web-monitoring-db/pull/237

> It would require just a tiny change to this line, no?
> [/src/components/search-bar/search-bar.jsx@`master`#L32](https://github.com/edgi-govdata-archiving/web-monitoring-ui/blob/master/src/components/search-bar/search-bar.jsx#L32) [](/edgi-govdata-archiving/web-monitoring-ui/blob/HEAD@{2019-10-22T02:16:27Z}/src/components/search-bar/search-bar.jsx#L32)

Yes and no. `*://*epa.gov/*` would match:

- `https://epa.gov/` (Yay!)
- `https://www.epa.gov/` (Yay!)
- `https://echo.epa.gov/` (Probably not what someone meant when typing `epa.gov` but maybe ok? Seems debatable, maybe problematic.)
- `https://deregulation_fun_at_the_epa.gov/` (This isn’t even a subdomain of `epa.gov`)

And if you did `*://*.epa.gov/*` you’d get `www.epa.gov` pages but not straight `epa.gov` pages.

Mr0grog commented:
Note also a lot of pages are at `www3.epa.gov`, not just `www.epa.gov`. (SURT handles this case well, but again, that field is not currently queryable.)

Mr0grog commented:
And if you want to explore the variety of subdomains and overlapping domain names: https://gist.github.com/Mr0grog/5e05ea795e7f6b2f4c20e97ff5a83cd9

SYU15 commented:
I see. Ok thanks for that background info, makes more sense. Seems like maybe it's best to keep it as is for now.

SYU15 commented:
Hey so I polished the PR up a bit and added some tests. I'm not quite satisfied with the test coverage, mainly because I think there should be tests for `web-monitoring-db` but that will take a bit of time for to scaffold from scratch (let me know if this is an issue). So might try to do that in a follow up PR unless there are some other features that are a higher priority.

Will also do some more manual testing to see if I can root out any other bugs/edge cases. And will fix whatever merge conflicts there are in a bit.

**Here is a screenshot for reference**
<img width="1211" alt="Screen Shot 2019-10-27 at 8 03 24 PM" src="https://user-images.githubusercontent.com/10149117/67649108-f2291c80-f8f4-11e9-8b33-cd4a05560137.png">

SYU15 commented on src/components/__tests__/search-bar-test.jsx:
```diff
@@ -16,20 +16,42 @@ describe('search-bar', () => {
+  it('Handles date range search queries for startDate', () => {
+    const onSearch = jest.fn();
+    const searchBar = mount(<SearchBar onSearch={onSearch} />);
+    searchBar.find('.DateInput_input.DateInput_input_1').first().simulate('focus');
```
I know these tests are brittle/annoying because they are relying on the classes from the library, so if you had any other suggestions as to how to handle this, lmk.

Mr0grog replied:
Yeah, I don’t think there are really better options here, so this is OK. Some techniques I use in this kind of situation:

- Lightweight: add a comment noting that this part is expected to be brittle, so it’s not totally confusing when it breaks for someone when updating the dependency. :)

- Midweight: wrap the logic of getting the input (or finding a day button, or whatever) so it’s more readable:

    ```js
    function getDateField (searchBar, inputType = 'FROM') {
      // NOTE: this is prone to break with updates to react-dates.
      const index = inputType === 'FROM' ? 0 : 1;
      return searchBar.find(`.DateInput_input.DateInput_input_1`).at(index);
    }

    it('does something I want to test', () => {
      const searchBar = mount(<SearchBar onSearch={onSearch} />);
      getDateField(searchBar, 'FROM').simulate('focus');
      getDateField(searchBar, 'TO').simulate('focus');
      expect(something);
    }
    ```

- Heavyweight: add a method to the `SearchBar` class to get the components of the calendar (you’d have to use refs for this). Almost certainly overkill in most cases.

Mr0grog replied:
OH! I totally missed that `<DateRangePicker>` requires you to specify IDs for these! Maybe you should just make those a class or instance property you can get at from the test? e.g:

```js
// Where you define the class, make the IDs something you can get:
class SearchDatePicker extends React.Component {
  static startInputId = 'startDate'
  static endInputId = 'endDate'

  render () {
    return (
      <DateRangePicker
        startDateId={SearchDatePicker + this.instanceId}
        endDateId={SearchDatePicker + this.instanceId}
        ... other props ...
      />
    );
  }
}

// In the test, use that!
function getDateField (searchBar, inputType = 'start') {
  // This is all public API and not brittle!
  const id = SearchDatePicker[`${inputType}DateId`];
  return searchBar.find(`#{id}`).first();
}
```

Mr0grog commented on src/components/search-bar/search-bar.jsx:
```diff
@@ -16,14 +18,34 @@ import './search-bar.css';
+        startDate: null,
+        endDate: null
+      }
+    };
```
I think it *probably* makes more sense to keep each of these as separate fields directly on the state and create the object when calling `onSearch`. That way there’s less to worry about with merging things and handling the state correctly.

Mr0grog commented on src/components/search-bar/search-bar.jsx:
```diff
@@ -16,14 +18,34 @@ import './search-bar.css';
     this._didSearch = this._didSearch.bind(this);
     this._dispatchSearch = debounce(this._dispatchSearch.bind(this), 500);
+    this.urlSearch = this.urlSearch.bind(this);
+    this.dateSearch = this.dateSearch.bind(this);
```
I’m OK if you don’t want to use `_` to prefix private methods, but we should stick to one style within this component. :)

Mr0grog commented on src/components/search-bar/search-bar.jsx:
```diff
@@ -16,14 +18,34 @@ import './search-bar.css';
+  componentDidUpdate(previousProps, previousState) {
+    if (previousState.query !== this.state.query) {
+      this.props.onSearch(this.state.query);
+    }
```
Depending how we manage state, this might not be an adequate check. (In the current setup, we should really be setting state with a function, and possibly merging objects there, which would mean we don’t always have a new object.)

Mr0grog commented on src/components/search-bar/search-bar.jsx:
```diff
@@ -16,14 +18,34 @@ import './search-bar.css';
+   *
+   * @param {String} url
+   */
   _dispatchSearch (url) {
```
The `dispatch` naming here existed because this is the function that *dispatches calls to `onSearch`*. Since that’s no longer true, this should probably be renamed. I’m thinking this function is actually what `urlSearch` should be.

(`_didSeach` is also no longer a very clear name now, either! Maybe `didChangeUrl` or `handleUrlInput` or something?)

Also, if you’re going to add doc comments here (and below, on `urlSearch`), they should include `@private` since this is a private function.

Mr0grog commented on src/components/search-bar/search-bar.jsx:
```diff
@@ -37,11 +59,31 @@ export default class SearchBar extends React.Component {
-      this.props.onSearch(query);
+      const query = url ? {url} : {url: null};
+      this.urlSearch(query);
     }
```
This probably doesn’t need the `if (this.props.onSearch)` guard anymore.

I’m also not sure it makes sense to make `urlSearch` a separate function from the code here, either. Maybe better to just collapse the two together?

Mr0grog commented on src/components/search-bar/search-bar.jsx:
```diff
@@ -37,11 +59,31 @@ export default class SearchBar extends React.Component {
+   */
+  urlSearch (urlQuery) {
+    const query = Object.assign({}, this.state.query, urlQuery);
+    this.setState({query});
```
If we are going to have a query object in the state (see comment in constructor), this should use an update function instead of merging ahead of time:

```js
this.setState(state => ({query: Object.assign(state.query, urlQuery)}));
```

Mr0grog commented on src/components/search-bar/search-bar.jsx:
```diff
@@ -37,11 +59,31 @@ export default class SearchBar extends React.Component {
+   */
+  dateSearch(dateQuery) {
+    const query = Object.assign({}, this.state.query, dateQuery);
+    this.setState({query});
```
Same as in `urlSearch` with how state is getting set.

Mr0grog commented on src/components/search-date-picker.jsx:
```diff
@@ -0,0 +1,58 @@
+ * @extends {React.Component}
+ * @param {SearchDatePickerProps} props
+ */
+export default class SearchDatePicker extends React.Component {
```
Hmmmm, I’m worried I might be missing the obvious, but is this component doing anything useful other than handling the focus state and providing some simple defaults?

If so, you could probably remove the dates from the state and simplify quite a bit:

```js
export default class SearchDatePicker extends React.Component {
  constructor (props) {
    super(props);
    this.state = { focusedInput: null };
  }

  // No componentDidUpdate or onDatesChange necessary.

  render() {
    return (
      <DateRangePicker
        startDateId="startDate"
        endDateId="endDate"
        startDate={this.props.startDate}
        endDate={this.props.endDate}
        onDatesChange={this.props.onDatesChange}
        focusedInput={this.state.focusedInput}
        onFocusChange={(focusedInput) => { this.setState({ focusedInput });}}
        isOutsideRange={day => !isInclusivelyBeforeDay(day, moment())}
        readOnly
        showClearDates />
    );
  }
}
```

This would require you to pass `startDate` and `endDate` in the props, but I’m pretty sure we’re goint to want to start doing that soon anyway (so we can restore the right value when switching between page list and details views).

Mr0grog commented on src/components/search-date-picker.jsx:
```diff
@@ -0,0 +1,58 @@
+    return (
+      <DateRangePicker
+        startDateId="startDate"
+        endDateId="endDate"
```
This is OK since we only ever use this component once, but this is probably not so great in the abstract case (e.g. maybe some future animated transition might put more than one of these in the DOM). It might be good to set some unique value in the constructor that we can add to these:

```js
  constructor (props) {
    // ...other constructor work...

    this.id = Math.random().toString().slice(2);
  }

  render () {
    return (
      <DateRangePicker
        startDateId={startDate + this.id}
        endDateId={endDate + this.id}
        ...other props...
      />
    )
  }
```

Or maybe this component could accept an `id` in its props, so you don’t need to gin one up in the constructor:

```js
  render () {
    return (
      <DateRangePicker
        startDateId={startDate + (this.props.id || '')}
        endDateId={endDate + (this.props.id || '')}
        ...other props...
      />
    )
  }
```

Mr0grog replied:
You might want to combine this with what I was thinking about the tests here: https://github.com/edgi-govdata-archiving/web-monitoring-ui/pull/422#discussion_r344904475

Mr0grog commented on src/components/web-monitoring-ui.jsx:
```diff
@@ -87,9 +105,11 @@ export default class WebMonitoringUi extends React.Component {
+        const query = Object.assign(
+          {include_latest: true},
+          this.state.query
+        );
```
No major concerns about reformatting this into multiple lines, but I *do* think it’s helpful to have the blank line between bailout/guard clauses like the `if (!loggedIn) return` above and the rest of the logic.

Mr0grog commented on src/services/utils/date-helpers.js:
```diff
@@ -0,0 +1,31 @@
+ */
+function _formatDate (date) {
+  if (!date) return '';
+  else if (moment.isMoment(date)) return date.utc().toISOString();
```
I probably wrote it this when suggesting, so my fault, but I think there’s no need for the `utc()` call here. The API will handle any time zone you send, and Moment’s `toISOString()` always returns a UTC-based timestamp anyway.

That also makes this the same as how we handle `Date` instances, we could just have one branch for both:

```js
  else if (moment.isMoment(date) || date instanceof Date) return date.toISOString();
```

Or maybe for compatibility with other date-like things:

```js
  else if (date.toISOString) return date.toISOString();
```

Mr0grog commented on src/services/utils/date-helpers.js:
```diff
@@ -0,0 +1,31 @@
+import moment from 'moment';
+
```
These kind of helpers are what `src/scripts` is for (yes it is TERRIBLY named), so this should probably be there instead of a new `src/servies/utils` directory.

On the other hand, all this logic is extremely specific to the `WebMonitoringDb` class. Is there a good reason to separate it into its own module?

Mr0grog commented on src/services/utils/date-helpers.js:
```diff
@@ -0,0 +1,31 @@
+ *
+ * @returns {String}
+ */
+function _formatDate (date) {
```
If this is meant to be private, it should be marked that way in the comment. Otherwise, let’s get rid of that `_`.

Mr0grog commented on src/services/web-monitoring-db.js:
```diff
@@ -287,6 +288,10 @@ export default class WebMonitoringDb {
         }
+        else if (key === CAPTURE_TIME) {
+          const formattedCaptureQuery = formatDateRange(value);
+          queryList.push(`${encodeURIComponent(key)}=${formattedCaptureQuery}`);
```
`formattedCaptureQuery` needs `encodeURIComponent`, too!

Mr0grog replied:
You could combine this with the below branch:

```js
if (value == null) {
  // whatever
}
else if (Array.isArray(value)) {
  // whatever
}
else {
  if (key === CAPTURE_TIME) {
    value = formatDateRange(value);
  }
  queryList.push(`${encodeURIComponent(key)}=${encodeURIComponent(value)}`);
}
```

Mr0grog commented on src/components/__tests__/search-bar-test.jsx:
```diff
@@ -16,20 +16,40 @@ describe('search-bar', () => {
   it('Handles search queries with a protocol correctly', () => {
-    const searchBar = shallow(<SearchBar onSearch={onSearch} />);
+    const onSearch = jest.fn();
+    const searchBar = shallow(<SearchBar onSearch={onSearch} startDate={null} endDate={null} />);
```
`startDate` and `endDate` don’t look like they are valid props on `SearchBar` (they *are* definitely on `SearchDatePicker`). Is this right?

SYU15 replied:
Bad copy/paste, my bad.

Mr0grog commented on src/components/__tests__/search-bar-test.jsx:
```diff
@@ -16,20 +16,40 @@ describe('search-bar', () => {
   it('Handles search queries without a protocol correctly', () => {
-    const searchBar = shallow(<SearchBar onSearch={onSearch} />);
+    const onSearch = jest.fn();
+    const searchBar = shallow(<SearchBar onSearch={onSearch} startDate={null} endDate={null} />);
```
Same as above w/r/t props.

Mr0grog commented on src/components/__tests__/search-bar-test.jsx:
```diff
@@ -16,20 +16,40 @@ describe('search-bar', () => {
+    searchBar.find('input#startDate1').simulate('focus');
+
+    searchBar.find('.CalendarDay.CalendarDay_1').first().simulate('click');
+    expect(onSearch.mock.calls[0][1]).toBeInstanceOf(moment);
```
If you want to be a bit more fancy, Jest gives you a way to check types when matching call arguments or objects:

```suggestion
    expect(onSearch).toHaveBeenCalledWith(null, expect.any(moment), null);
```

**But that’s only very slightly clearer than what you already have, so ¯\\\_(ツ)\_/¯**

(Same for the other test below.)

SYU15 commented on src/components/search-bar/search-bar.jsx:
```diff
@@ -36,20 +76,34 @@ export default class SearchBar extends React.Component {
+   */
+  dateSearch ({ startDate, endDate }) {
+    this.setState({ startDate, endDate });
+  }
```
Nope, this is passed into the `SearchDatePicker` component.

Mr0grog replied:
Yeah, but it’s still not public — it’s in charge of passing the handle around, something else can’t come get it from a search bar instance. That’s what I was trying to say.

Mr0grog replied:
Anyway, not really a big deal.

SYU15 replied:
Ok updated it.

SYU15 commented on src/scripts/__tests__/db-helpers-test.js:
```diff
@@ -0,0 +1,27 @@
+    const dateRangeString = formatDateRange({startDate: '', endDate: ''});
+
+    expect(dateRangeString).toBe('..');
+  });
```
Hmm...good point. Will remove it.

SYU15 commented on src/components/web-monitoring-ui.jsx:
```diff
@@ -65,11 +65,27 @@ export default class WebMonitoringUi extends React.Component {
+    }
+
+    this.setState({ query: formattedQuery });
     this.loadPages();
```
I'll remove this if the search loader gets checked in first.

SYU15 commented:
> Very minor and pedantic note: a lot of your doc comments could use periods at the end of the sentences. 😬

I added a bunch of periods!

SYU15 commented:
> Sorry about having been so slow on basically all reviews for the last couple weeks. I know the need for cleanup and conflict resolution here is basically my fault.

No problem. I've rebased/tested and updated the PR. Do you think it's good to merge?

Mr0grog commented:
👍 Seems good to me!
